enhancements/etcd/scaling-etcd-with-raft-learners.md

```diff
@@ -0,0 +1,343 @@
+2. `ETCD_MEMBER_IDS`: This variable will be added to the etcd static pod EMV, the mapping of this data enables extra
+precautions during scaling.
+
+- Writer: new function added to `envVarFns`[1].
+
+- Consumer: cluster-etcd-operator verify membership command.
+
+New etcd members on an existing node will need to archive the data-dir of the previous member to allow scale up to
+occur. `verify membership` can observe this by asking the cluster for membership details and using the conditions of
+`Name=""` and the method of `Member` `IsLearner` to conclude this.
+
+[1] https://github.com/openshift/cluster-etcd-operator/blob/release-4.10/pkg/etcdenvvar/etcd_env.go#L43
+
+### Cluster Member Controller
+
+The cluster member controller manages scaling up etcd during member replacement and bootstrap. This proposal makes
+changes to the way this controller manages scale up.
+
+`ensureEtcdLearnerPromotion`: This new method will provide the logic necessary to ask the cluster for the list of
+members and attempt to promote any learner members which have a log in sync with the leader.
+
+#### Reads
+
+1. openshift-etcd pods/quorum-guard: Today the controller loops not `Ready` etcd pods and will attempt to scale
+up if the etcd is not already part of the cluster membership. While this is effective during bootstrap it is not
+efficient as scaling no longer needs to be serial. Also by ensuring the quorum guard pod exists on the node before
+scale up we can give more power to the PDB in scaling.
+
+2. `MemberListResponse`: The Cluster API `MemberListRequest` RPC provides `[]Member` for the etcd cluster including the
+method `IsLearner`.
+
+- Consumers: scale up and promotion of learner members
+
+- ref: https://pkg.go.dev/github.com/coreos/etcd/etcdserver/etcdserverpb#MemberListResponse
+
+3. `MemberPromoteResponse`: The Cluster API `MemberPromoteRequest` asks the leader of the term if the learner's log
+is in sync with its own. If no return error.
+
+- Consumers `ensureEtcdLearnerPromotion`
+
+- ref: https://pkg.go.dev/go.etcd.io/etcd/api/v3/etcdserverpb#MemberPromoteResponse
+
+### Risks and Mitigations
+
+The largest risks to scaling is quorum loss, data loss, and split brain scenario.
+
+- **Quorum Loss**: Because new members are added as non voting members and can not be promoted to voting members unless
+the etcd process starts and the log has been completely replicated and in sync with leader.
+
+
+- **Split brain**: The cluster gates starting of the etcd process on a verification process based on the cluster
+member id. This ensures that each revision has an explicit expected membership. Because quorum guard replicas are
+managed by the operator the cluster topology will remain in a safe configuration (odd number of members).
+
+
+- **Data loss**: A concern with rolling scaling of etcd with large data files is possible data loss. The MTTR of log
+replication is directly tied to the size of the state. If members of the cluster were replaced too quickly in
```

Comment:
> The MTTR of log replication is directly tied to the size of the state.

If you happen to have any figures that show etcd size vs MTTR, this would be useful context for understanding this risk